src/rustc-driver-example.rs

```diff
@@ -0,0 +1,103 @@
+#![feature(rustc_private)]
+
+extern crate rustc;
+extern crate rustc_error_codes;
+extern crate rustc_errors;
+extern crate rustc_hash;
+extern crate rustc_hir;
+extern crate rustc_interface;
+extern crate rustc_span;
+
+use rustc::session;
+use rustc::session::config;
+use rustc_errors::registry;
+use rustc_hash::{FxHashMap, FxHashSet};
+use rustc_interface::interface;
+use rustc_span::source_map;
+use std::path;
+use std::process;
+use std::str;
+
+fn main() {
+    let out = process::Command::new("rustc")
+        .arg("--print=sysroot")
+        .current_dir(".")
+        .output()
+        .unwrap();
+    let sysroot = str::from_utf8(&out.stdout).unwrap().trim();
+    let filename = "main.rs";
+    let contents = "static HELLO: &str = \"Hello, world!\"; fn main() { println!(\"{}\", HELLO); }";
+    let errors = registry::Registry::new(&rustc_error_codes::DIAGNOSTICS);
+    let config = interface::Config {
+        // Command line options
+        opts: config::Options {
+            maybe_sysroot: Some(path::PathBuf::from(sysroot)),
+            ..config::Options::default()
+        },
+
+        // cfg! configuration in addition to the default ones
+        // FxHashSet<(String, Option<String>)>
+        crate_cfg: FxHashSet::default(),
+
+        input: config::Input::Str {
+            name: source_map::FileName::Custom(String::from(filename)),
+            input: String::from(contents),
+        },
+        // Option<PathBuf>
+        input_path: None,
+        // Option<PathBuf>
+        output_dir: None,
+        // Option<PathBuf>
```

Comment:
Not sure listing the types is very helpful.

Reply:
If you know nothing about the compiler internals (like me!) it's quite helpful for understanding what each of these options actually does.